### ISSUE TITLE
[Feature Request] Implementation of cosine_distance_with_negative_samples loss

### ISSUE THREAD

**Issue by iamsimha** [tensorflow/tensorflow#19159]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:1.5.0
- **Python version**: 
- **Bazel version (if compiling from source)**:No
- **GCC/Compiler version (if compiling from source)**:No
- **CUDA/cuDNN version**:CPU
- **GPU model and memory**:CPU
- **Exact command to reproduce**:Feature request

### Describe the problem
Tensorflow current does not have [CNTK's negative sampling loss](https://www.cntk.ai/pythondocs/cntk.losses.html#cntk.losses.cosine_distance_with_negative_samples). I tried to get the same functionality using tf.nn.sampled_softmax_loss, but the weights parameter of the function needs to be variable, instead of an op[If weights is an arbitrary op, I get sparse gradient warning].

Is it possible to implement this loss natively without doing an embedding lookup on weights?

**Comment by martinwicke** [comment]
@drpngx do you know whether this is possible (efficiently)?

**Comment by drpngx** [comment]
As I understand it, it looks similar to the sampled softmax, except that the negative samples are sampled deterministically using a stride operation.

@iamsimha can you confirm?

**Comment by iamsimha** [comment]
The sampled_softmax_loss is different from CNTKs negative sampling loss in two ways.

1. The negative samples are generated by probabilistic sampling in TF, instead of deterministic strides.

2. The weights parameter to sampled_softmax_loss can only be a tf.variable instead of an arbitrary op. This is because sampled_softmax_loss internally does a nn.embedding_lookup on weights parameter. According to my understanding embedding_lookup is efficient, only if the input is a variable or an op with sparse gradient support. If the input parameter is an arbitrary op, TF raises a sparse gradient warning.  So is it possible to support arbitrary ops for the weights parameter by removing dependency on embedding_lookup?

**Comment by martinwicke** [comment]
Re 2: The warning is just a warning -- it means that whatever you're doing on the way from the embedding lookup to the underlying variables might requires a dense gradient. If that is the case, then that may be inefficient. It does not mean that it is necessarily inefficient. You can check the graph to see whether in your case you actually instantiate a large dense gradient. We're open to changing the warning to be clearer.

Re 1: We would prefer not to add a new function for this. If this is a critical use case (I am not sure it is, would deterministic, but pseudorandom sampling be enough?) then we should extend the interface of the existing method to support it.
